src/Symfony/Component/DependencyInjection/Tests/Compiler/IntegrationTest.php

```diff
@@ -113,4 +113,30 @@ public function testProcessInlinesWhenThereAreMultipleReferencesButFromTheSameDe
         $this->assertFalse($container->hasDefinition('b'));
         $this->assertFalse($container->hasDefinition('c'), 'Service C was not inlined.');
     }
+
+    /**
+     * @expectedException \Symfony\Component\DependencyInjection\Exception\ServiceCircularReferenceException
+     */
+    public function testCircularReferencesCausedByMethodCallsAreDetectedDuringCompilation()
+    {
+        $container = new ContainerBuilder();
+        $container->setResourceTracking(false);
+
+        $container
+            ->register('foobar', '\stdClass')
+            ->addArgument(new Reference('foo'))
+        ;
+
+        $container
+            ->register('foo', '\stdClass')
+            ->addArgument(new Reference('bar'))
+        ;
+
+        $container
+            ->register('foo', '\stdClass')
```

Comment:
Well, a circular reference is still  a circular reference. I just wanted to explicitly have one with an intermediate service.